controllers/datadogagent/component/agent/new.go

```diff
@@ -0,0 +1,91 @@
+
+package agent
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/DataDog/datadog-operator/controllers/datadogagent/component"
+
+	edsv1alpha1 "github.com/DataDog/extendeddaemonset/api/v1alpha1"
+)
+
+// NewDaemonset use to generate the skeleton of a new daemonset based on few information
+func NewDaemonset(owner metav1.Object, componentKind, componentName, version string, selector *metav1.LabelSelector) *appsv1.DaemonSet {
+	labels, annotations, selector := component.GetDefaultMetadata(owner, componentKind, componentName, version, selector)
+
+	daemonset := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        componentName,
+			Namespace:   owner.GetNamespace(),
+			Labels:      labels,
+			Annotations: annotations,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: selector,
+		},
+	}
+	return daemonset
+}
+
+// NewExtendedDaemonset use to generate the skeleton of a new extended daemonset based on few information
+func NewExtendedDaemonset(owner metav1.Object, edsOptions *ExtendedDaemonsetOptions, componentKind, componentName, version string, selector *metav1.LabelSelector) *edsv1alpha1.ExtendedDaemonSet {
+	// FIXME (@CharlyF): The EDS controller uses the Spec.Selector as a node selector to get the NodeList to rollout the agent.
+	// Per https://github.com/DataDog/extendeddaemonset/blob/28a8e082cee9890ae6d925a7d6247a36c6f6ba5d/controllers/extendeddaemonsetreplicaset/controller.go#L344-L360
+	// Up until v0.8.2, the Datadog Operator set the selector to nil, which circumvented this case.
+	// Until the EDS controller uses the Affinity field to get the NodeList instead of Spec.Selector, let's keep the previous behavior.
+	labels, annotations, _ := component.GetDefaultMetadata(owner, componentKind, componentName, version, selector)
+
+	daemonset := &edsv1alpha1.ExtendedDaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        componentName,
+			Namespace:   owner.GetNamespace(),
+			Labels:      labels,
+			Annotations: annotations,
+		},
+		Spec: defaultEDSSpec(edsOptions),
+	}
+
+	return daemonset
+}
+
+func defaultEDSSpec(options *ExtendedDaemonsetOptions) edsv1alpha1.ExtendedDaemonSetSpec {
+	spec := edsv1alpha1.ExtendedDaemonSetSpec{}
+	edsv1alpha1.DefaultExtendedDaemonSetSpec(&spec, edsv1alpha1.ExtendedDaemonSetSpecStrategyCanaryValidationModeAuto)
+
+	if options.MaxPodUnavailable != "" {
+		spec.Strategy.RollingUpdate.MaxUnavailable = newIntOrStringPointer(options.MaxPodUnavailable)
+	}
+
+	if options.MaxPodSchedulerFailure != "" {
+		spec.Strategy.RollingUpdate.MaxPodSchedulerFailure = newIntOrStringPointer(options.MaxPodSchedulerFailure)
+	}
+
+	if options.CanaryDuration != 0 {
+		spec.Strategy.Canary.Duration = &metav1.Duration{Duration: options.CanaryDuration}
+	}
+
+	if options.CanaryPodCount != "" {
+		spec.Strategy.Canary.Replicas = newIntOrStringPointer(options.CanaryPodCount)
```

Comment:
what's the reason for the parameter rename? is it planned to change this in the EDS as well?

Reply:
I think choose this name without checking what was the option name in EDS. 
I think make more sens now to name it:  `options.CanaryReplicas` WDYT?

Reply:
sounds good, thanks